File: substrate/source_attach_settings_decision_mo_compose.py
"""HTML-native source attach over settings decision + MO unattended pack.

remote_fetched / pdf_primary always False.
live_router_authorized / live_execution_authorized always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.html_native_source_attach_compose import (
    HtmlNativeSourceAttachCompose,
    HtmlNativeSourceAttachComposeError,
    compose_html_native_source_attach,
)
from substrate.settings_decision_mo_unattended_fullscreen_compose import (
    SettingsDecisionMoUnattendedFullscreenCompose,
    SettingsDecisionMoUnattendedFullscreenComposeError,
    compose_settings_decision_mo_unattended_fullscreen,
)


class SourceAttachSettingsDecisionMoComposeError(ValueError):
    """Fail-closed validation for source attach + settings decision MO pack."""


@dataclass(frozen=True)
class SourceAttachSettingsDecisionMoCompose:
    session_id: str
    parent_asset_id: str
    title: str
    account_id: str
    week_id: str
    asset_id: str
    sources: HtmlNativeSourceAttachCompose
    settings_mo: SettingsDecisionMoUnattendedFullscreenCompose
    pack_ready: bool
    remote_fetched: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    store_mutated: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    live_execution_authorized: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    draft_written: bool
    analysis_written: bool
    purchase_executed: bool
    charge_executed: bool
    hosted: bool
    twin_written: bool
    prompts_injected: bool
    live_dispatch_authorized: bool
    backlog_mutated: bool
    suite_rewritten: bool
    remote_index_queried: bool
    inventory_mutated: bool
    record_persisted: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "sources": self.sources.to_dict(),
            "settings_mo": self.settings_mo.to_dict(),
            "pack_ready": self.pack_ready,
            "remote_fetched": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "store_mutated": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "live_execution_authorized": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "draft_written": False,
            "analysis_written": False,
            "purchase_executed": False,
            "charge_executed": False,
            "hosted": False,
            "twin_written": False,
            "prompts_injected": False,
            "live_dispatch_authorized": False,
            "backlog_mutated": False,
            "suite_rewritten": False,
            "remote_index_queried": False,
            "inventory_mutated": False,
            "record_persisted": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": "source_attach_settings_decision_mo_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise SourceAttachSettingsDecisionMoComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_source_attach_settings_decision_mo(
    *,
    sources: object,
    settings_mo: object,
    operator_ack: object,
    require_both: object | None = None,
) -> SourceAttachSettingsDecisionMoCompose:
    """Source attach + settings decision MO. Never remote-fetches."""
    if not isinstance(operator_ack, bool):
        raise SourceAttachSettingsDecisionMoComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(sources, dict):
        raise SourceAttachSettingsDecisionMoComposeError(
            "sources must be an object"
        )
    if not isinstance(settings_mo, dict):
        raise SourceAttachSettingsDecisionMoComposeError(
            "settings_mo must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise SourceAttachSettingsDecisionMoComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "remote_fetched=false · pdf_primary=false · store_mutated=false",
        "live_router_authorized=false · live_execution_authorized=false",
        "production_router_verdict=REJECT",
    ]

    try:
        src = compose_html_native_source_attach(
            session_id=sources.get("session_id"),
            parent_asset_id=sources.get("parent_asset_id"),
            requested_families=sources.get("requested_families"),
            sources=sources.get("sources"),
            operator_ack=operator_ack,
        )
    except HtmlNativeSourceAttachComposeError as e:
        raise SourceAttachSettingsDecisionMoComposeError(str(e)) from e
    notes.extend(f"[sources] {n}" for n in src.notes)

    try:
        sm = compose_settings_decision_mo_unattended_fullscreen(
            decision=settings_mo.get("decision"),
            mo_pack=settings_mo.get("mo_pack"),
            operator_ack=operator_ack,
            require_both=settings_mo.get("require_both"),
        )
    except SettingsDecisionMoUnattendedFullscreenComposeError as e:
        raise SourceAttachSettingsDecisionMoComposeError(str(e)) from e
    notes.extend(f"[settings_mo] {n}" for n in sm.notes)

    session = _require_nonempty(src.session_id, field="session_id")
    parent = _require_nonempty(src.parent_asset_id, field="parent_asset_id")
    title = _require_nonempty(sm.title, field="title")
    account = _require_nonempty(sm.account_id, field="account_id")
    week = _require_nonempty(sm.week_id, field="week_id")
    asset = _require_nonempty(sm.asset_id, field="asset_id")

    session_aligned = sm.session_id == session
    parent_aligned = sm.parent_asset_id == parent
    if not session_aligned:
        notes.append(
            "session_id mismatch between sources and settings_mo — pack_ready blocked"
        )
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between sources and settings_mo — "
            "pack_ready blocked"
        )

    if require:
        pack_ready = (
            session_aligned
            and parent_aligned
            and src.attach_ready is True
            and sm.pack_ready is True
            and src.remote_fetched is False
            and src.pdf_view_authorized is False
            and src.store_mutated is False
            and sm.live_router_authorized is False
            and sm.live_execution_authorized is False
            and sm.purchase_executed is False
            and sm.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned
            and parent_aligned
            and operator_ack is True
            and src.remote_fetched is False
            and sm.purchase_executed is False
            and sm.production_router_verdict == "REJECT"
            and (src.attach_ready is True or sm.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — source attach + settings decision MO pack ready; "
            "still pure"
        )
    else:
        notes.append(
            "pack_ready=false — sources, settings_mo, alignment, or operator_ack "
            "gate open"
        )

    if (
        src.remote_fetched is not False
        or src.pdf_view_authorized is not False
        or src.store_mutated is not False
        or sm.live_router_authorized is not False
        or sm.live_execution_authorized is not False
        or sm.purchase_executed is not False
        or sm.production_router_verdict != "REJECT"
    ):
        raise SourceAttachSettingsDecisionMoComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "remote_fetched=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "store_mutated=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "live_execution_authorized=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "draft_written=false",
            "analysis_written=false",
            "purchase_executed=false",
            "charge_executed=false",
            "hosted=false",
            "twin_written=false",
            "prompts_injected=false",
            "live_dispatch_authorized=false",
            "backlog_mutated=false",
            "suite_rewritten=false",
            "remote_index_queried=false",
            "inventory_mutated=false",
            "record_persisted=false",
            "production_router_verdict=REJECT",
        )
    )

    return SourceAttachSettingsDecisionMoCompose(
        session_id=session,
        parent_asset_id=parent,
        title=title,
        account_id=account,
        week_id=week,
        asset_id=asset,
        sources=src,
        settings_mo=sm,
        pack_ready=pack_ready,
        remote_fetched=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        store_mutated=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        live_execution_authorized=False,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        draft_written=False,
        analysis_written=False,
        purchase_executed=False,
        charge_executed=False,
        hosted=False,
        twin_written=False,
        prompts_injected=False,
        live_dispatch_authorized=False,
        backlog_mutated=False,
        suite_rewritten=False,
        remote_index_queried=False,
        inventory_mutated=False,
        record_persisted=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority="source_attach_settings_decision_mo_compose_advisory",
    )


def format_source_attach_settings_decision_mo_summary(
    c: SourceAttachSettingsDecisionMoCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"attach_ready={c.sources.attach_ready} · "
        f"html_ready={c.sources.html_ready_count} · "
        f"settings_mo_ready={c.settings_mo.pack_ready} · "
        f"verdict={c.production_router_verdict} · "
        f"remote_fetched=false · live_router_authorized=false · "
        f"live_execution_authorized=false"
    )


__all__ = [
    "SourceAttachSettingsDecisionMoCompose",
    "SourceAttachSettingsDecisionMoComposeError",
    "compose_source_attach_settings_decision_mo",
    "format_source_attach_settings_decision_mo_summary",
]

File: apps/reading/src/api/sourceAttachSettingsDecisionMoCompose.ts
/**
 * HTML-native arxiv/substack source attach over settings decision-tree +
 * Midnight Oil unattended fullscreen pack (pure).
 *
 * Operator vision: call knowledge-dense publications into deep research while
 * choosing the model driver with budget bar/projection and MO unattended
 * readiness — every source HTML-native, never remote-fetched in pure layer.
 *
 * remote_fetched / pdf_primary always false.
 * live_router_authorized / live_execution_authorized always false.
 * production_router_verdict always REJECT.
 */

import {
  composeHtmlNativeSourceAttach,
  type HtmlNativeSourceAttachCompose,
  type HtmlNativeSourceAttachInput,
} from "./htmlNativeSourceAttachCompose";
import {
  composeSettingsDecisionMoUnattendedFullscreen,
  type SettingsDecisionMoUnattendedFullscreenCompose,
  type SettingsDecisionMoUnattendedFullscreenInput,
} from "./settingsDecisionMoUnattendedFullscreenCompose";

export interface SourceAttachSettingsDecisionMoInput {
  sources: Omit<HtmlNativeSourceAttachInput, "operator_ack">;
  settings_mo: Omit<
    SettingsDecisionMoUnattendedFullscreenInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface SourceAttachSettingsDecisionMoCompose {
  session_id: string;
  parent_asset_id: string;
  title: string;
  account_id: string;
  week_id: string;
  asset_id: string;
  sources: HtmlNativeSourceAttachCompose;
  settings_mo: SettingsDecisionMoUnattendedFullscreenCompose;
  pack_ready: boolean;
  remote_fetched: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  store_mutated: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  live_execution_authorized: false;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  draft_written: false;
  analysis_written: false;
  purchase_executed: false;
  charge_executed: false;
  hosted: false;
  twin_written: false;
  prompts_injected: false;
  live_dispatch_authorized: false;
  backlog_mutated: false;
  suite_rewritten: false;
  remote_index_queried: false;
  inventory_mutated: false;
  record_persisted: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "source_attach_settings_decision_mo_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Source attach + settings decision + MO unattended fullscreen.
 * Never remote-fetches; never live-routes; ND REJECT.
 */
export function composeSourceAttachSettingsDecisionMo(
  input: SourceAttachSettingsDecisionMoInput,
): SourceAttachSettingsDecisionMoCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.sources || typeof input.sources !== "object") {
    throw new Error("sources must be an object");
  }
  if (!input.settings_mo || typeof input.settings_mo !== "object") {
    throw new Error("settings_mo must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "remote_fetched=false · pdf_primary=false · store_mutated=false",
    "live_router_authorized=false · live_execution_authorized=false",
    "production_router_verdict=REJECT",
  ];

  const sources = composeHtmlNativeSourceAttach({
    ...input.sources,
    operator_ack: input.operator_ack,
  });
  notes.push(...sources.notes.map((n) => `[sources] ${n}`));

  const settings_mo = composeSettingsDecisionMoUnattendedFullscreen({
    ...input.settings_mo,
    operator_ack: input.operator_ack,
  });
  notes.push(...settings_mo.notes.map((n) => `[settings_mo] ${n}`));

  const session_id = requireNonEmpty(sources.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    sources.parent_asset_id,
    "parent_asset_id",
  );
  const title = requireNonEmpty(settings_mo.title, "title");
  const account_id = requireNonEmpty(settings_mo.account_id, "account_id");
  const week_id = requireNonEmpty(settings_mo.week_id, "week_id");
  const asset_id = requireNonEmpty(settings_mo.asset_id, "asset_id");

  const session_aligned = settings_mo.session_id === session_id;
  const parent_aligned = settings_mo.parent_asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between sources and settings_mo — pack_ready blocked",
    );
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between sources and settings_mo — pack_ready blocked",
    );
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      sources.attach_ready === true &&
      settings_mo.pack_ready === true &&
      sources.remote_fetched === false &&
      sources.pdf_view_authorized === false &&
      sources.store_mutated === false &&
      settings_mo.live_router_authorized === false &&
      settings_mo.live_execution_authorized === false &&
      settings_mo.purchase_executed === false &&
      settings_mo.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      sources.remote_fetched === false &&
      settings_mo.purchase_executed === false &&
      settings_mo.production_router_verdict === "REJECT" &&
      (sources.attach_ready === true || settings_mo.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — source attach + settings decision MO pack ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — sources, settings_mo, alignment, or operator_ack gate open",
    );
  }

  if (
    sources.remote_fetched !== false ||
    sources.pdf_view_authorized !== false ||
    sources.store_mutated !== false ||
    settings_mo.live_router_authorized !== false ||
    settings_mo.live_execution_authorized !== false ||
    settings_mo.purchase_executed !== false ||
    settings_mo.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("remote_fetched=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("store_mutated=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("live_execution_authorized=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("purchase_executed=false");
  notes.push("charge_executed=false");
  notes.push("hosted=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("backlog_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("remote_index_queried=false");
  notes.push("inventory_mutated=false");
  notes.push("record_persisted=false");
  notes.push("production_router_verdict=REJECT");

  return {
    session_id,
    parent_asset_id,
    title,
    account_id,
    week_id,
    asset_id,
    sources,
    settings_mo,
    pack_ready,
    remote_fetched: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    store_mutated: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    live_execution_authorized: false,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    draft_written: false,
    analysis_written: false,
    purchase_executed: false,
    charge_executed: false,
    hosted: false,
    twin_written: false,
    prompts_injected: false,
    live_dispatch_authorized: false,
    backlog_mutated: false,
    suite_rewritten: false,
    remote_index_queried: false,
    inventory_mutated: false,
    record_persisted: false,
    production_router_verdict: "REJECT",
    notes,
    authority: "source_attach_settings_decision_mo_compose_advisory",
  };
}

export function formatSourceAttachSettingsDecisionMoSummary(
  c: SourceAttachSettingsDecisionMoCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `attach_ready=${c.sources.attach_ready} · ` +
    `html_ready=${c.sources.html_ready_count} · ` +
    `settings_mo_ready=${c.settings_mo.pack_ready} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `remote_fetched=false · live_router_authorized=false · live_execution_authorized=false`
  );
}
